internal/handlers/common/projection.go

```diff
@@ -221,28 +221,264 @@ func projectDocumentWithoutID(doc *types.Document, projection *types.Document, i
 			)
 
 		case *types.Array, string, types.Binary, types.ObjectID,
-			time.Time, types.NullType, types.Regex, types.Timestamp: // all this types are treated as new fields value
+			time.Time, types.NullType, types.Regex, types.Timestamp: // all these types are treated as new fields value
 			projected.Set(key, value)
 
 		case bool: // field: bool
-			// process top level fields
-			if path.Len() == 1 {
-				if inclusion {
-					if docWithoutID.Has(key) {
-						projected.Set(key, must.NotFail(docWithoutID.Get(key)))
-					}
-
-					continue
+			if inclusion {
+				// inclusion projection copies the field on the path from docWithoutID to projected.
+				if _, err = includeProjection(path, docWithoutID, projected); err != nil {
+					return nil, err
 				}
 
-				projected.Remove(key)
+				continue
 			}
 
-			// TODO: process dot notation here https://github.com/FerretDB/FerretDB/issues/2430
+			// exclusion projection removes the field on the path in projected.
+			excludeProjection(path, projected)
 		default:
 			return nil, lazyerrors.Errorf("unsupported operation %s %v (%T)", key, value, value)
 		}
 	}
 
 	return projected, nil
 }
+
+// pathNotExists indicates that no path was found.
+var pathNotExists = errors.New("path not exists")
+
+// includeProjection copies the field on the path from source to projected.
+// When an array is on the path, it returns the array containing any document
+// with the same key. Dot notation with array index path does not include
+// the field unlike document.SetByPath(path).
+// Inclusion projection with non-existent path creates an empty document
```

Comment:
From this description, I don't really understand in which cases this function returns errors. Based on this description, a non-existent path doesn't look like an error situation to me.